src/prototype_room_market.js

```diff
@@ -32,49 +30,93 @@ Room.prototype.trySellOrders = function(resource, sellAmount) {
   return sellAmount;
 };
 
-Room.prototype.handleMarket = function() {
-  if (!this.terminal) {
-    return false;
+Room.prototype.sellByOthersOrders = function(sellAmount, resource) {
+  const sortByEnergyCostAndPrice = order => Game.market.calcTransactionCost(sellAmount, this.name, order.roomName) +
+    -order.price * sellAmount / config.market.energyCreditEquivalent;
+  if (Memory.orders[ORDER_BUY][resource]) {
+    const orders = _.sortBy(Memory.orders[ORDER_BUY][resource].orders, sortByEnergyCostAndPrice);
+    for (let order of orders) {
+      const amount = Math.min(sellAmount, order.remainingAmount);
+      if (amount > 0 && order.price >= config.market.minSellPrice) {
+        if (Game.market.calcTransactionCost(amount, this.name, order.roomName) > this.terminal.store.energy) {
+          break;
+        }
+        this.log(order.id, this.name, amount, order.price);
+        let returnCode = Game.market.deal(order.id, amount, this.name);
+        this.log('market.deal:', resource, returnCode);
+        if (returnCode === OK) {
+          break;
+        }
+      }
+    }
   }
+};
 
+Room.prototype.sellOwnMineral = function() {
   const minerals = this.find(FIND_MINERALS);
   const resource = minerals[0].mineralType;
 
   if (!this.terminal.store[resource]) {
     return false;
   }
 
-  if (this.terminal.store[resource] <= config.market.minAmount) {
+  if (this.terminal.store[resource] <= config.market.minAmountToSell) {
     return false;
   }
 
-  let sellAmount = this.terminal.store[resource] - config.market.minAmount;
+  let sellAmount = this.terminal.store[resource] - config.market.minAmountToSell;
 
-  if (config.market.trySellOrders) {
-    sellAmount = this.trySellOrders(resource, sellAmount);
+  if (config.market.sellByOwnOrders) {
+    sellAmount = this.sellByOwnOrders(resource, sellAmount);
   }
 
   if (sellAmount <= 0) {
     return true;
   }
+  this.sellByOthersOrders(sellAmount, resource);
+  return true;
+};
 
-  const sortByEnergyCostAndPrice = order => Game.market.calcTransactionCost(sellAmount, this.name, order.roomName) +
-    -order.price * sellAmount / config.market.energyCreditEquivalent;
+Room.prototype.buyByOthersOrders = function(resource) {
+  const avgBuyPrice = brain.getMarketOrderAverage(ORDER_BUY, resource);
+  const avgSellPrice = brain.getMarketOrderAverage(ORDER_SELL, resource);
+  const myBuyPrice = Math.min((avgBuyPrice || 1) * config.market.buyOrderPriceMultiplicator, config.market.maxBuyPrice);
+  let buyAmount = this.terminal.store[resource] || config.market.maxAmountToBuy;
```

Comment:
``let buyAmount = config.market.maxAmountToBuy - (this.terminal.store[resource] || 0)``